packages/devtools/lib/src/timeline/flame_chart_canvas.dart

```diff
@@ -481,35 +482,64 @@ class FlameChartNode {
         ..stroke();
     }
 
-    canvas
-      ..fillStyle = colorToCss(textColor)
-      ..font = fontStyleToCss(const TextStyle(fontSize: _fontSize));
-
-    // TODO(kenzie): polish this. Sometimes we trim excessively. We should do
-    // something smarter here to be more exact. 'm' is arbitrary - it was
-    // selected after trial and error with different letter measurements.
-    // Measure the text and truncate based on the actual width of the available
-    // area.
-    final singleLetterWidth = canvas.measureText('m').width;
-    final maxLetters = math.min(
-      math.max(0, (rect.width - horizontalPadding * 2) ~/ singleLetterWidth),
-      selected ? text.length : math.min(text.length, maxDisplayTextLength),
-    );
+    if (rect.width > minimumRectWidthForText) {
+      canvas
+        ..fillStyle = colorToCss(selected ? selectedTextColor : textColor)
+        ..font = fontStyleToCss(const TextStyle(fontSize: _fontSize));
+
+      String displayText = text;
+
+      // TODO(kenzie): further optimize text painting by setting a budget for
+      // how much text can be measured each frame, and incrementally updating
+      // the text in the UI.
+      if (!_textFitsInRect(displayText, canvas)) {
+        displayText = longestFittingSubstring(text, maxTextWidth, canvas);
+      }
 
-    String displayText = text.substring(0, maxLetters);
-    if (maxLetters == maxDisplayTextLength &&
-        text.length > maxDisplayTextLength) {
-      displayText += '...';
+      canvas.fillText(
+        displayText,
+        rect.left + horizontalPadding,
+        rect.top + _textOffsetY,
+        maxTextWidth,
+      );
     }
+  }
 
-    final maxTextWidth = rect.width - horizontalPadding * 2;
+  bool _textFitsInRect(String text, CanvasRenderingContext2D canvas) {
+    final textWidth = textMeasurements[text] ??= canvas.measureText(text).width;
+    return textWidth <= maxTextWidth;
+  }
 
-    canvas.fillText(
-      displayText,
-      rect.left + horizontalPadding,
-      rect.top + _textOffsetY,
-      maxTextWidth,
-    );
+  String longestFittingSubstring(
+    String originalText,
+    num maxWidth,
+    CanvasRenderingContext2D canvas,
+  ) {
+    final firstLetter = originalText.substring(0, 1);
+    final firstLetterWidth =
+        textMeasurements[firstLetter] ??= canvas.measureText(firstLetter).width;
+    if (firstLetterWidth >= maxWidth) {
+      return '';
+    }
+
+    int binarySearchForLongestFittingIndex() {
```

Comment:
there are more efficient ways to do this. Add a TODO to optimize this method.

Reply:
also add a test for the text measuring logic so when you optimize it, things still work correctly.